apps/public/public_test.go

```diff
@@ -237,6 +237,80 @@ func TestAcquireTokenWithTenantID(t *testing.T) {
 	}
 }
 
+func TestWithInstanceDiscovery(t *testing.T) {
+	// replacing browserOpenURL with a fake for the duration of this test enables testing AcquireTokenInteractive
+	realBrowserOpenURL := browserOpenURL
+	defer func() { browserOpenURL = realBrowserOpenURL }()
+	browserOpenURL = fakeBrowserOpenURL
+
+	accessToken := "*"
+	clientInfo := base64.RawStdEncoding.EncodeToString([]byte(`{"uid":"uid","utid":"utid"}`))
+	stackurl := "stack.local"
+	host := fmt.Sprintf("https://%s/", stackurl)
+
+	for _, tenant := range []string{
+		"adfs",
+		"98b8267d-e97f-426e-8b3f-7956511fd63f",
+	} {
+		for _, method := range []string{"authcode", "devicecode", "interactive", "password"} {
+			t.Run(method, func(t *testing.T) {
+				authority := host + tenant
+				mockClient := mock.Client{}
+				mockClient.AppendResponse(mock.WithBody(mock.GetTenantDiscoveryBody(stackurl, tenant)))
+				if method == "devicecode" {
+					mockClient.AppendResponse(mock.WithBody([]byte(`{"device_code":"...","expires_in":600}`)))
+				} else if method == "password" && tenant != "adfs" {
+					// user realm metadata, which is not requested when AuthorityType is ADFS
+					mockClient.AppendResponse(mock.WithBody([]byte(`{"account_type":"Managed","cloud_audience_urn":"urn","cloud_instance_name":"...","domain_name":"..."}`)))
+				}
+				mockClient.AppendResponse(
+					mock.WithBody(mock.GetAccessTokenBody(accessToken, mock.GetIDToken(tenant, authority), "rt", clientInfo, 3600)),
+				)
+				client, err := New("client-id", WithAuthority(authority), WithHTTPClient(&mockClient), WithInstanceDiscovery(false))
+				if err != nil {
+					t.Fatal(err)
+				}
+				ctx := context.Background()
+				if _, err = client.AcquireTokenSilent(ctx, tokenScope); err == nil {
+					t.Fatal("silent auth should fail because the cache is empty")
+				}
+
+				var ar AuthResult
+				var dc DeviceCode
+				switch method {
+				case "authcode":
+					ar, err = client.AcquireTokenByAuthCode(ctx, "auth code", "https://localhost", tokenScope)
+				case "devicecode":
+					dc, err = client.AcquireTokenByDeviceCode(ctx, tokenScope)
+				case "interactive":
+					ar, err = client.AcquireTokenInteractive(ctx, tokenScope)
+				case "password":
+					ar, err = client.AcquireTokenByUsernamePassword(ctx, tokenScope, "username", "password")
+				default:
+					t.Fatal("test bug: no test for " + method)
+				}
+				if err != nil {
+					t.Fatal(err)
+				}
+				if method == "devicecode" {
+					if ar, err = dc.AuthenticationResult(ctx); err != nil {
+						t.Fatal(err)
+					}
+				}
+				if ar.AccessToken != accessToken {
```

Comment:
I think 2 extra things need to be checked: 

1. HomeAccountId property of Account objects from should be in format "oid.oid" (for AAD it's "oid.tid" but we don't have "tid" in ADFS). This is pretty important, otherwise other MSALs might not be able to read from the cache.
2. Account object (which is populated from the id token) should have the username set. I believe ADFS will expose this in the "upn" claim and AAD in the "preferred_username" claim. IMO, I would set Username to `idToken.preferred_username ?? idToken.upn`

Reply:
CC @chlowell @rayluo

Reply:
We are setting HomeAccountID like this currently. Do we have to make changes to this? And is this something that should go through in this pr? 

// HomeAccountID creates the home account ID.
func (c ClientInfo) HomeAccountID() string {
	if c.UID == "" || c.UTID == "" {
		return ""
	}
	return fmt.Sprintf("%s.%s", c.UID, c.UTID)
}

Reply:
I think so, otherwise you risk introducing a security issue or a cache miss issue. If home account Id is "", MSAL might not be able to retrieve the correct token from the cache (or it might not be able to retrieve one at all). 

if think the logic is: 

- if both are emtpy, homeAccountId is empty (this is probably for client_credentials)
- if c.Uitd == "", then use `fmt.Sprintf("%s.%s", c.UID, c.UID)`

Reply:
> 2 extra things need to be checked:
>
> * HomeAccountId property of Account objects from should be in format "oid.oid" (for AAD it's "oid.tid" but we don't have "tid" in ADFS). This is pretty important, otherwise other MSALs might not be able to read from the cache.

AFAIK, this is an undefined corner case in the original Token Cache Schema. Existing MSALs probably already have inconsistent behaviors here, ranging from "oid"-only to "oid.adfs". The effect is indeed different MSALs would not hit the same account in the cache. Not sure whether it is prevalent in the wild. We did not hear report on this. It is ADFS, anyway.

But if MSAL Go currently returns a completely empty string as HomeAccountID, that sounds too extreme. Better still return something, at least the oid.

> * Account object (which is populated from the id token) should have the username set. I believe ADFS will expose this in the "upn" claim and AAD in the "preferred_username" claim. IMO, I would set Username to `idToken.preferred_username ?? idToken.upn`

Agreed. [MSAL Python did it the same way](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/1.20.0/msal/token_cache.py#L15-L18), too.

> And is this something that should go through in this pr?

Not necessarily in this PR, but if it would break Azure Stack scenarios, we may need to fix it sooner or later.

Reply:
Then there would be cases where we would have UID.UID in AAD, when we want "" right? Does this pose some type of security issue for AAD? If so, sounds like we should distinguish between ADFS and AAD in the function/make a function for each auth type.

Reply:
In AAD, the client info has `uid` and `utid` claims. In ADFS, it only has `uid` claim. You don't need to know if authority is ADFS to see that the client_info does not have `utid` claim. 

Just avoid setting HomeAccountId to "" if the `uid` claim is present. This is a potential security issue, because the cache logic will not be able to differentiate between users.